### PR TITLE
866 Callback Urls

### PR DESCRIPTION
**Pull request by Ennosigaeon**
**Pull Request checklist**

- [ ] The commit(s) message(s) follows the contribution [guidelines](doc/contribution-guidelines.md#commit-message-format) ?
- [ ] Tests for the changes have been added (for bug fixes / features) ?
- [ ] Docs have been added / updated (for bug fixes / features) ?

**Current behavior :** See #866 

**New behavior :** An optional `callbackUrl` parameter can be provided when starting a (async/sync) job. Once the job is finished (or fails), the result is pushed to the callback url via an HTTP Post request. The content of the Post request is identical to the output produced by sync jobs. Currently, the callbacks are only available for jobs but maybe a similar behaviour can also be later added for contexts is desired.

**Other information**: For local testing, I have started netcat in an additonal shell to simulate a simple HTTP server
```
while true; do echo -e "HTTP/1.1 200 OK\r\n\r\nOk" | nc -l -q 1 -p 6666; done
```
The output of each job with a callback (e.g. `curl -d "input.string = a b c a b see" "localhost:8090/jobs?appName=test&classPath=spark.jobserver.MyErrorJob&callbackUrl=http%3A%2F%2Flocalhost%3A6666"`) is logged in netcat.

---
This change is [Reviewable](https://reviewable.io/reviews/spark-jobserver/spark-jobserver/1339)

### CODE REVIEW

File: job-server/src/test/scala/spark/jobserver/JobManagerActorSpec.scala [hunk 1]
```diff
@@ -242,6 +247,7 @@ class JobManagerTestActor(daoActor: ActorRef,
   }
 }
 
+// TODO test callback
```
[the remark this answers is not in the document]
Comment (Ennosigaeon) [reply job-server/src/test/scala/spark/jobserver/JobManagerActorSpec.scala]: Gone now

File: job-server/src/main/resources/db/h2/migration/V1_1__add_callback_url.sql [hunk 2]
```diff
@@ -0,0 +1 @@
+ALTER TABLE JOBS ADD COLUMN CALLBACK_URL VARCHAR(255);
```
[the remark this answers is not in the document]
Comment (Ennosigaeon) [reply job-server/src/main/resources/db/h2/migration/V1_1__add_callback_url.sql]: Renamed to `V3__add...`

File: job-server/src/main/resources/db/mysql/migration/V1_1__add_callback_url.sql [hunk 3]
```diff
@@ -0,0 +1 @@
+ALTER TABLE `JOBS` ADD COLUMN `CALLBACK_URL` VARCHAR(255);
```
[the remark this answers is not in the document]
Comment (Ennosigaeon) [reply job-server/src/main/resources/db/mysql/migration/V1_1__add_callback_url.sql]: Renamed to `V2__add...`

File: job-server/src/main/resources/db/postgresql/migration/V1_1__add_callback_url.sql [hunk 4]
```diff
@@ -0,0 +1 @@
+ALTER TABLE "JOBS" ADD COLUMN "CALLBACK_URL" VARCHAR(255);
```
[the remark this answers is not in the document]
Comment (Ennosigaeon) [reply job-server/src/main/resources/db/postgresql/migration/V1_1__add_callback_url.sql]: Renamed to `V2__add...`

**Comment by valan4ik** [comment]
@Ennosigaeon,
the change works for me in a spark client mode locally but doesn't work in a "cluster" mode. Is it a problem with my setup or do you observe the same behavior?

Please rebase the change.

**Comment by Ennosigaeon** [comment]
Ok, I think I found the issue. I missed to adapt the `StartJobSerializer`. Should be fine now.

**Review by valan4ik — APPROVED** [review]
LGTM. Change worked locally, I also executed partially integration tests.

**Comment by Ennosigaeon** [comment]
I have squashed the commits and rebased on master

**Comment by Ennosigaeon** [comment]
Any updates on this PR?

**Review by valan4ik — APPROVED** [review]

**Comment by valan4ik** [comment]
@Ennosigaeon, could you please resolve the conflicts in WebApi imports?
I will merge the PR right after

**Comment by Ennosigaeon** [comment]
@valan4ik: I have rebased the PR on master and resolved the conflict.

**Review by valan4ik — APPROVED** [review]

**Pull request merged**